espnet2/layers/augmentation.py

```diff
@@ -0,0 +1,535 @@
+        n_fft (int): length of FFT (in second)
+        win_length (int or None): The window length (in second) used for STFT
+            If None, it is treated as equal to n_fft
+        hop_length (int): The hop size (in second) used for STFT
+        window (str or None): The windowing function applied to the signal after
+            padding with zeros
+
+    Returns:
+        ret (torch.Tensor): perturbed signal (..., time)
+    """
+    n_fft = int(sample_rate * n_fft)
+    if hop_length is None:
+        hop_length = n_fft // 4
+    else:
+        hop_length = int(sample_rate * hop_length)
+    if win_length is None:
+        win_length = n_fft
+    if window is not None:
+        window_func = getattr(torch, f"{window}_window")
+        window = window_func(win_length, dtype=waveform.dtype, device=waveform.device)
+    spec = torch.stft(
+        waveform, n_fft, hop_length, win_length, window=window, return_complex=True
+    )
+    freq = spec.size(-2)
+    phase_advance = torch.linspace(0, math.pi * hop_length, freq)[..., None]
+    spec_sp = torchaudio.functional.phase_vocoder(spec, factor, phase_advance)
+    len_stretch = int(round(waveform.size(-1) / factor))
+    ret = torch.functional.istft(
+        spec_sp, n_fft, hop_length, win_length, window=window, length=len_stretch
+    )
+    return ret
+
+
+def codecs(
+    waveform,
+    sample_rate: int,
+    format: str,
+    compression: Optional[float] = None,
+    encoding: Optional[str] = None,
+    bits_per_sample: Optional[int] = None,
+):
+    """Apply the specified codecs to the input signal.
+
+    Warning: Wait until torchaudio 2.1 for this function to work.
+
+    Note:
+        1. This function only supports CPU backend.
+        2. The GSM codec can be used to emulate phone line channel effects.
+
+    Args:
+        waveform (torch.Tensor): audio signal (..., time)
+        sample_rate (int): sampling rate in Hz
+        format (str): file format.
+            Valid values are "wav", "mp3", "ogg", "vorbis", "amr-nb", "amb",
+            "flac", "sph", "gsm", and "htk".
+        compression (float or None, optional): used for formats other than WAV
+
+            For more details see torchaudio.backend.sox_io_backend.save().
+        encoding (str or None, optional): change the encoding for the supported formats
+            Valid values are "PCM_S" (signed integer Linear PCM),
+            "PCM_U" (unsigned integer Linear PCM), "PCM_F" (floating point PCM),
+            "ULAW" (mu-law), and "ALAW" (a-law).
+            For more details see torchaudio.backend.sox_io_backend.save().
+        bits_per_sample (int or None, optional): change the bit depth
+            for the supported formats
+            For more details see torchaudio.backend.sox_io_backend.save().
+
+    Returns:
+        ret (torch.Tensor): compressed signal (..., time)
+    """
+    ret = torchaudio.functional.apply_codec(
```

Comment:
how about adding some warning or exception to not be called for unwanted torch version? (or in a different place because if you put that here, it can be called too often)

Reply:
For now, I think I can just raise NotImplementedError for this function.